File: backend/app/agent/sanitize.py
"""
Artifact sanitization.
Defense-in-depth: server-side bleach + documented allowlist.
See architecture.md for full security rationale.

Allowed HTML tags and attributes:
  Structural: p, div, section, article, header, footer, main, aside, nav
  Headings:   h1, h2, h3, h4, h5, h6
  Lists:      ul, ol, li, dl, dt, dd
  Inline:     span, strong, em, b, i, u, s, del, ins, mark, small, sub, sup
  Code:       code, pre, kbd, samp, var
  Links:      a (href, title only — target forced to _blank, rel=noopener)
  Quotes:     blockquote, cite, q
  Tables:     table, thead, tbody, tfoot, tr, th, td, caption, colgroup, col
  Media:      (none — img, video, audio are blocked to prevent data exfil)
  Other:      br, hr

Blocked unconditionally:
  script, style, iframe, object, embed, form, input, button, select, textarea,
  link, meta, base, noscript, template, slot, canvas, svg, math
  All event attributes (onclick, onload, onerror, etc.)
  javascript: and data: URIs in href/src
"""
import re
import bleach
from bleach.css_sanitizer import CSSSanitizer

# ─── Allowlists ───────────────────────────────────────────────────────────────

ALLOWED_TAGS = [
    # Structure
    "p", "div", "section", "article", "header", "footer", "main", "aside", "nav",
    # Headings
    "h1", "h2", "h3", "h4", "h5", "h6",
    # Lists
    "ul", "ol", "li", "dl", "dt", "dd",
    # Inline
    "span", "strong", "em", "b", "i", "u", "s", "del", "ins", "mark", "small", "sub", "sup",
    # Code
    "code", "pre", "kbd", "samp", "var",
    # Links
    "a",
    # Quotes
    "blockquote", "cite", "q",
    # Tables
    "table", "thead", "tbody", "tfoot", "tr", "th", "td", "caption", "colgroup", "col",
    # Misc
    "br", "hr",
]

ALLOWED_ATTRIBUTES = {
    "a": ["href", "title"],
    "th": ["scope", "colspan", "rowspan"],
    "td": ["colspan", "rowspan"],
    "blockquote": ["cite"],
    "code": ["class"],  # for syntax highlighting class names only
    "pre": ["class"],
}

# Block javascript: and data: URIs
def _clean_link(attrs, new=False):
    href = attrs.get((None, "href"), "")
    if href.lower().startswith(("javascript:", "data:", "vbscript:")):
        return None  # drop the attribute
    # Force safe external links
    attrs[(None, "target")] = "_blank"
    attrs[(None, "rel")] = "noopener noreferrer"
    return attrs


def sanitize_html(raw_html: str) -> str:
    """
    Sanitize HTML content for safe rendering in the Artifact Viewer.
    Strips all disallowed tags, event attributes, and dangerous URIs.
    """
    cleaned = bleach.clean(
        raw_html,
        tags=ALLOWED_TAGS,
        attributes=ALLOWED_ATTRIBUTES,
        strip=True,        # strip disallowed tags (not escape them)
        strip_comments=True,
    )
    # Apply link cleaner to handle javascript: hrefs that bleach may miss
    cleaned = bleach.linkify(
        cleaned,
        callbacks=[_clean_link],
        skip_tags=["pre", "code"],
    )
    return cleaned


def sanitize_markdown_to_text(markdown: str) -> str:
    """
    For Markdown artifacts: return as-is (client renders via marked + DOMPurify).
    Light server-side check: block embedded HTML script tags.
    """
    # Remove any <script> blocks that might be in the markdown
    cleaned = re.sub(r"<script[^>]*>.*?</script>", "", markdown, flags=re.DOTALL | re.IGNORECASE)
    cleaned = re.sub(r"<style[^>]*>.*?</style>", "", cleaned, flags=re.DOTALL | re.IGNORECASE)
    return cleaned

File: frontend/src/lib/sanitize.ts
/**
 * Client-side Markdown rendering with DOMPurify sanitization.
 * HTML artifacts are rendered in sandboxed iframes (see ArtifactViewer).
 *
 * What this sanitizer PERMITS:
 *   - All standard Markdown elements (headings, lists, code blocks, tables, links)
 *   - Bold, italic, inline code, blockquotes
 *   - HTML embedded in Markdown that passes DOMPurify allowlist
 *
 * What it BLOCKS:
 *   - <script> tags and all JS event handlers
 *   - <iframe>, <object>, <embed>, <form>
 *   - javascript: and data: URIs
 *   - Anything not in DOMPurify's default safe list
 *
 * For HTML artifacts, the iframe sandbox attribute blocks:
 *   - Script execution (no allow-scripts)
 *   - Navigation (no allow-top-navigation)
 *   - Popups (no allow-popups)
 *   - Form submission (no allow-forms)
 * See ArtifactViewer.tsx for the iframe implementation.
 */
import { marked } from 'marked';
import DOMPurify from 'dompurify';

// Configure marked for better output
marked.setOptions({
  gfm: true,
  breaks: true,
});

// DOMPurify config — strict allowlist
const PURIFY_CONFIG: DOMPurify.Config = {
  ALLOWED_TAGS: [
    'p', 'div', 'span', 'section', 'article', 'header', 'footer', 'main',
    'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
    'ul', 'ol', 'li', 'dl', 'dt', 'dd',
    'strong', 'em', 'b', 'i', 'u', 's', 'del', 'ins', 'mark', 'small', 'sub', 'sup',
    'code', 'pre', 'kbd', 'samp',
    'a', 'blockquote', 'cite', 'q',
    'table', 'thead', 'tbody', 'tfoot', 'tr', 'th', 'td', 'caption',
    'br', 'hr',
  ],
  ALLOWED_ATTR: ['href', 'title', 'class', 'colspan', 'rowspan', 'scope'],
  FORBID_TAGS: ['script', 'style', 'iframe', 'object', 'embed', 'form', 'input', 'button'],
  FORBID_ATTR: ['onerror', 'onload', 'onclick', 'onmouseover', 'onfocus', 'onblur'],
  ALLOW_DATA_ATTR: false,
};

export function renderMarkdown(markdown: string): string {
  const rawHtml = marked.parse(markdown) as string;
  return DOMPurify.sanitize(rawHtml, PURIFY_CONFIG);
}

export function sanitizeHtml(html: string): string {
  return DOMPurify.sanitize(html, PURIFY_CONFIG);
}
